docs/adr/0003-workspace-blob-caching.md

```diff
@@ -0,0 +1,56 @@
+# 3. Workspace BLOB Caching
+
+* Status: [ **proposed** | rejected | accepted | deprecated ]
+* Date: 2023-09-20 
+* Authors: @chanwit
+* Deciders: TBD 
+
+## Context
+
+The TF-Controller currently faces challenges related to the deletion of Terraform resources.
+These problems span across three categories:
+
+1. Single object deletion,
+2. Resources with dependencies deletion, and
+3. Namespace deletion.
+
+These problems must be fixed in the above order as (2) and (3) require single object deletion to be resolved first.
+
+Deleting a single TF object can sometimes be obstructed because it's tied to other resources like Source objects, Secrets, and ConfigMaps. If we try to remove it without deleting these resources, the TF object gets stuck in an inconsistent state, making it harder for users to manage their infrastructure smoothly.
+Therefore, the TF-Controller is being enhanced to address this problem more efficiently, using the contents of generated Workspace BLOBs. Each BLOB contains all necessary information from the associated Source, Secrets, and ConfigMaps to ensure that TF-Controller finalization procedures can delete objects correctly.
+
+Currently, the TF-Controller downloads a Source BLOB and pushes it to a tf-runner. The tf-runner processes this BLOB to create a Workspace file system. It generates a backend configuration file, variable files, and other necessary files for the Workspace file system, using data from associated Secrets and ConfigMaps. This newly created Workspace file system is then compressed, sent back to the TF-Controller, and stored as a Workspace BLOB in the controller's storage.
```

Comment:
>  This newly created Workspace file system is then compressed, sent back to the TF-Controller, and stored as a Workspace BLOB in the controller's storage.

This is given as part of the solution below -- "A gRPC function named `CreateWorkspaceBlob` will be invoked by the TF-Controller to tell tf-runner to compress the Workspace file system into a tar.gz BLOB, which is then retrieved back to the controller". Is it current behaviour, or proposed behaviour?